### PR TITLE
Sanjo/vault release

### PR DESCRIPTION
**Pull request by SanjoDeundiak**

### CODE REVIEW

**Review by mikelodder7 — APPROVED** [review]
LGTM

**Review by jared-s — APPROVED** [review]
Since you were curious about coverage, I ran it manually. Looks like good coverage on the impls.

```
Filename                                                                                                                                 Regions    Missed Regions     Cover   Functions  Missed Functions  Executed       Lines      Missed Lines     Cover
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
/ockam/ockam_core/src/error.rs                                                                                                                 4                 3    25.00%           2                 1    50.00%          16                13    18.75%
/ockam/ockam_vault_core/src/macros.rs                                                                                                          1                 0   100.00%           1                 0   100.00%           1                 0   100.00%
/ockam/ockam_vault_core/src/secret.rs                                                                                                          3                 1    66.67%           2                 1    50.00%           7                 1    85.71%
/ockam/ockam_vault_core/src/types.rs                                                                                                          31                23    25.81%           7                 0   100.00%          19                 4    78.95%
/rustc/368275062fb655c1f36e0398f88b15379a1f3c93/library/std/src/thread/local.rs                                                                2                 0   100.00%           2                 0   100.00%           2                 0   100.00%
src/error.rs                                                                                                                                   4                 3    25.00%           1                 1     0.00%           4                 1    75.00%
src/hash_impl.rs                                                                                                                              55                13    76.36%           4                 0   100.00%          90                 8    91.11%
src/kid_impl.rs                                                                                                                               22                 3    86.36%           4                 0   100.00%          33                 1    96.97%
src/lib.rs                                                                                                                                     1                 0   100.00%           1                 0   100.00%           1                 0   100.00%
src/secret_impl.rs                                                                                                                            64                11    82.81%           8                 0   100.00%         115                 8    93.04%
src/signer_impl.rs                                                                                                                            16                 2    87.50%           2                 0   100.00%          30                 1    96.67%
src/software_vault.rs                                                                                                                         27                 8    70.37%          10                 1    90.00%          47                 2    95.74%
src/verifier_impl.rs                                                                                                                           6                 2    66.67%           1                 0   100.00%          14                 2    85.71%
src/xeddsa.rs                                                                                                                                 11                 0   100.00%           3                 0   100.00%          67                 0   100.00%```

**Comment by jared-s** [comment]
Let's rename the  `KidVault` trait to `KeyIdVault` and everywhere rename `kid` to `key_id`

**Comment by jared-s** [comment]
An update to the README and `lib.rs` rustdoc header:


```Various Ockam protocols, like secure channels, key lifecycle, credential exchange, device enrollment etc. depend on a variety of standard cryptographic primitives or building blocks. Depending on the environment, these building blocks may be provided by a software implementation or a cryptographically capable hardware component.

In order to support a variety of cryptographically capable hardware we maintain loose coupling between our protocols and how a specific building block is invoked in a specific hardware. This is achieved using an abstract Vault trait.

A concrete implementation of the Vault trait is called an Ockam Vault. Over time, and with help from the Ockam open source community, we plan to add vaults for several TEEs, TPMs, HSMs, and Secure Enclaves.

This crate provides a software-only Vault implementation that can be used when no cryptographic hardware is available. The primary Ockam crate uses this as the default Vault implementation. 
```

Based on: https://github.com/ockam-network/ockam/blob/develop/documentation/concepts/vaults.md

**Comment by SanjoDeundiak** [comment]
> Let's rename the `KidVault` trait to `KeyIdVault` and everywhere rename `kid` to `key_id`

I think KID is generally accepted form of KeyId, still we can rename. @mikelodder7 @mrinalwadhwa thoughts?

File: implementations/rust/ockam/ockam_vault/src/secret_impl.rs [hunk 1]
```diff
@@ -1,36 +1,38 @@
...
+    fn secret_generate(&mut self, attributes: SecretAttributes) -> ockam_core::Result<Secret> {
         let mut rng = OsRng {};
-        let key = match attributes.stype {
+        let (key, key_id) = match attributes.stype {
             SecretType::Curve25519 => {
                 let sk = x25519_dalek::StaticSecret::new(&mut rng);
-                SecretKey::new(sk.to_bytes().to_vec())
+                let public = x25519_dalek::PublicKey::from(&sk);
+                let private = SecretKey::new(sk.to_bytes().to_vec());
+                let key_id = self.sha256(public.as_bytes())?;
```
Comment (jared-s) [inline implementations/rust/ockam/ockam_vault/src/secret_impl.rs:21]: We need to expand all uses of `?` and `unwrap` to `match` or `if let`s throughout the Vault code

File: implementations/rust/ockam/ockam_vault/src/hash_impl.rs [hunk 2]
```diff
@@ -1,32 +1,54 @@
...
+
         // FIXME: Doesn't work for secrets with size more than 32 bytes
         let okm_len = output_attributes.len() * 32;
 
         let okm = {
             let mut okm = vec![0u8; okm_len];
             let prk = hkdf::Hkdf::<Sha256>::new(Some(salt.key().as_ref()), ikm);
             prk.expand(info, okm.as_mut_slice())
-                .or(Err(Error::HkdfExpandError.into()))?;
+                .or(Err(VaultError::HkdfExpandError.into()))?;
```
Comment (jared-s) [inline implementations/rust/ockam/ockam_vault/src/hash_impl.rs:51]: `clippy` is mad at this line, can you run it and fix up the warnings? The other warnings are similar in other areas

**Review by jared-s — CHANGES_REQUESTED** [review]

File: implementations/rust/ockam/ockam_vault/src/secret_impl.rs (hunk 1 above)
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_vault/src/secret_impl.rs:21]: @jared-s all the unwraps in the vault are inside tests. Except one unwrap in xeddsa @mikelodder7 
And what's wrong with `?`

File: implementations/rust/ockam/ockam_vault/src/lib.rs [hunk 3]
```diff
@@ -1,10 +1,22 @@
+//! Software implementation of ockam_vault_core traits.
+//!
+//! This crate contains one of the possible implementation of the vault traits
+//! which you can use with Ockam library.
+
```
[the remark this answers is not in the document]
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_vault/src/lib.rs:5]: @jared-s have problem with this piece of code after enabling all the warnings
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=343a801fee8b013be66e2bed6f872002

File: implementations/rust/ockam/ockam_vault/src/hash_impl.rs (hunk 2 above)
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_vault/src/hash_impl.rs:51]: @jared-s this becomes ugly https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=472d31cb68cf7ca636d43350f4dbfbc8

File: implementations/rust/ockam/ockam_vault/src/secret_impl.rs (hunk 1 above)
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_vault/src/secret_impl.rs:21]: @mikelodder7 fixed the unwrap

**Review by jared-s — APPROVED** [review]

**Comment by jared-s** [comment]
Thanks for all of those fixes @SanjoDeundiak, after we land https://github.com/ockam-network/ockam/pull/845 can you please rebase and then we will land this PR?

File: implementations/rust/ockam/ockam_vault/Cargo.toml [hunk 4]
```diff
@@ -1,26 +1,32 @@
 [package]
-name = "ockam_vault"
-version = "0.0.0"
 authors = ["Ockam Developers"]
+description = """
```
Comment (jared-s) [inline implementations/rust/ockam/ockam_vault/Cargo.toml:3]:
```suggestion
description = """ A software-only Ockam Vault implementation that can be used when no cryptographic hardware is available.
```

File: implementations/rust/ockam/ockam_vault_core/Cargo.toml [hunk 5]
```diff
@@ -1,15 +1,24 @@
 [package]
-name = "ockam_vault_core"
-version = "0.0.0"
 authors = ["Ockam Developers"]
+description = """
```
Comment (jared-s) [inline implementations/rust/ockam/ockam_vault_core/Cargo.toml:3]:
```suggestion
description = """The abstract Ockam Vault trait.
```

**Comment by jared-s** [comment]
@SanjoDeundiak looking great. Could you please squash/reword any commit description of "kid vault" so that it doesn't appear anywhere in the source history?

**Pull request merged**